src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -269,6 +268,35 @@ private void registerExtendedClass(String qualifiedClassName,
         }
     }
 
+    /**
+     * Calculates and returns the type declaration name matching count.
+     *
+     * <p>
+     * Suppose our pattern class is {@code foo.a.b} and class to be matched is
+     * {@code foo.a.ball} then type declaration name matching count would be calculated by
+     * comparing every character, and updating main counter when we hit "." to prevent matching
+     * "a.b" with "a.ball". In this case type declaration name matching count
+     * would be equal to 6 and not 7 (b of ball is not counted).
+     * </p>
+     *
+     * @param patternClass class against which the given class has to be matched
+     * @param classToBeMatched class to be matched
+     * @return class name matching count
+     */
+    private static int typeDeclarationNameMatchingCount(String patternClass,
```

Comment:
Ok, both this method and `getAnonSuperTypeMatchingCount` are hard to understand, and bring up a lot of questions (why prefix ++ sometimes, etc.). Additionally, when we find the next bug, it is tempting to extend logic within the loops, so my concern is also for maintainability.

Is there a reason that we cannot use a more general approach to both of these methods? Something like:
1. Turn both strings into lists of type Character
2. Take intersection of lists
3. Remove everything after after final `.`
4. Apply any special logic
5. Return count of remaining characters

1-3 could certainly be extracted to some util method like 'String getIntersection(String x, String y)`. Then, each one of these "specialty" methods can apply custom logic and return a count.